parquet/src/main/java/org/apache/iceberg/parquet/TypeWithSchemaVisitor.java

```diff
@@ -19,187 +19,50 @@
-                    // key result remains null
-                  }
-                  break;
-                default:
-                  // both results will remain null
-              }
-
-              return visitor.map(map, group, keyResult, valueResult);
-
-            } finally {
-              visitor.fieldNames.pop();
-            }
-
-          default:
-        }
-      }
-
-      Types.StructType struct = iType != null ? iType.asStructType() : null;
-      return visitor.struct(struct, group, visitFields(struct, group, visitor));
+  @Override
+  protected Type arrayElementType(Type arrayType) {
+    if (arrayType == null) {
+      return null;
     }
-  }
 
-  private static <T> T visitField(Types.NestedField iField, Type field, TypeWithSchemaVisitor<T> visitor) {
-    visitor.fieldNames.push(field.getName());
-    try {
-      return visit(iField != null ? iField.type() : null, field, visitor);
-    } finally {
-      visitor.fieldNames.pop();
-    }
+    return arrayType.asListType().elementType();
   }
 
-  private static <T> List<T> visitFields(Types.StructType struct, GroupType group, TypeWithSchemaVisitor<T> visitor) {
-    List<T> results = Lists.newArrayListWithExpectedSize(group.getFieldCount());
-    for (Type field : group.getFields()) {
-      int id = -1;
-      if (field.getId() != null) {
-        id = field.getId().intValue();
-      }
-      Types.NestedField iField = (struct != null && id >= 0) ? struct.field(id) : null;
-      results.add(visitField(iField, field, visitor));
+  @Override
+  protected Type mapKeyType(Type mapType) {
+    if (mapType == null) {
+      return null;
     }
 
-    return results;
-  }
-
-  public T message(Types.StructType iStruct, MessageType message, List<T> fields) {
-    return null;
-  }
-
-  public T struct(Types.StructType iStruct, GroupType struct, List<T> fields) {
-    return null;
+    return mapType.asMapType().keyType();
   }
 
-  public T list(Types.ListType iList, GroupType array, T element) {
-    return null;
-  }
-
-  public T map(Types.MapType iMap, GroupType map, T key, T value) {
-    return null;
-  }
+  @Override
+  protected Type mapValueType(Type mapType) {
+    if (mapType == null) {
+      return null;
+    }
 
-  public T primitive(org.apache.iceberg.types.Type.PrimitiveType iPrimitive,
-                     PrimitiveType primitive) {
-    return null;
+    return mapType.asMapType().valueType();
   }
 
-  protected String[] currentPath() {
-    return Lists.newArrayList(fieldNames.descendingIterator()).toArray(new String[0]);
-  }
+  @Override
+  protected Type fieldType(Type structType, int pos, Integer fieldId) {
+    if (structType == null || fieldId == null) {
+      return null;
+    }
 
-  protected String[] path(String name) {
-    List<String> list = Lists.newArrayList(fieldNames.descendingIterator());
-    list.add(name);
-    return list.toArray(new String[0]);
+    Types.NestedField field = structType.asStructType().field(fieldId);
```

Comment:
I think it's a bug here ?  Because in the method definition, we have a `parquetFieldId` but here we use this id to access the iceberg nested field ? That sounds unreasonable.

```java
  protected abstract P fieldType(P structType, int pos, Integer parquetFieldId);
```

Reply:
btw: we might need to have a unit test to cover this case ?